src/sema/contracts.rs

```diff
@@ -733,6 +733,39 @@ fn check_inheritance(contract_no: usize, ns: &mut ast::Namespace) {
     ns.diagnostics.extend(diagnostics);
 }
 
+/// Given a contract number, check for function names conflicting with any mangled name.
+/// Only applies to public functions.
+///
+/// Note: In sema we do not care about the function name too much.
+/// The mangled name is consumed later by the ABI generation.
+pub fn mangle_function_names(contract_no: usize, ns: &mut ast::Namespace) {
+    let public_functions: Vec<usize> = ns.contracts[contract_no]
+        .all_functions
+        .keys()
+        .copied()
+        .filter(|f| ns.functions[*f].is_public() && ns.functions[*f].ty == pt::FunctionTy::Function)
+        .collect();
+
+    for f in &public_functions {
+        if let Some(offender) = public_functions
+            .iter()
+            .find(|other| ns.functions[*f].mangled_name == ns.functions[**other].name)
+        {
+            let f = &ns.functions[*f];
+            let message = format!(
+                "mangling the symbol of overloaded function '{}' with signature '{}' results in a new symbol '{}' but this symbol already exists",
+                &f.name, &f.signature, &f.mangled_name
+            );
+            ns.diagnostics.push(ast::Diagnostic::error_with_note(
+                f.loc,
+                message,
+                ns.functions[*offender].loc,
+                "this function declaration conflicts with mangled name".into(),
+            ))
+        }
+    }
```

Comment:
We should have a test for when a function in a base contract conflicts with a mangled name,

Reply:
there is, see `tests/contract_testcases/substrate/functions/mangling_01.dot`